Enable system power management on mcxn236

ZhaoxiangJin opened the pull request:
Enable system power management on mcxn236.
depends on:


- [x] https://github.com/zephyrproject-rtos/zephyr/pull/99360
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/99877
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/99862
- [x] https://github.com/zephyrproject-rtos/zephyr/pull/99784


test with power_mgmt_soc:
<details>
<summary>Logs</summary>

```
*** Booting Zephyr OS build v4.3.0-1132-gde1c93fd9c49 ***
Running TESTSUITE power_mgmt
===================================================================
START - test_pm_dummyinit
I: PM dummy single-thread test started for one cycle
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 0.013 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 1.184 seconds
I: PM dummy single-thread completed
 PASS - test_pm_dummyinit in 1.217 seconds
===================================================================
START - test_pm_multithread
I: Thread task A init
I: Thread task B init
I: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 0.013 seconds
I: Resume
I: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 1.184 seconds
I: Resume
I: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 0.013 seconds
I: Resume
I: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 1.184 seconds
I: Resume
I: PM multi-thread completed
I: PM state[0] entry counter 0

I: PM state[0] exit counter 0

I: PM state[1] entry counter 4

I: PM state[1] exit counter 4

I: PM state[2] entry counter 0

I: PM state[2] exit counter 0

I: PM state[3] entry counter 4

I: PM state[3] exit counter 4

I: PM state[4] entry counter 0

I: PM state[4] exit counter 0

I: PM state[5] entry counter 0

I: PM state[5] exit counter 0

I: PM state[6] entry counter 0

I: PM state[6] exit counter 0

 PASS - test_pm_multithread in 2.479 seconds
===================================================================
START - test_pm_singlethread
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 0.013 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 1.184 seconds
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 0.013 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 1.184 seconds
I: PM single-thread completed
I: PM state[0] entry counter 0

I: PM state[0] exit counter 0

I: PM state[1] entry counter 3

I: PM state[1] exit counter 3

I: PM state[2] entry counter 0

I: PM state[2] exit counter 0

I: PM state[3] entry counter 4

I: PM state[3] exit counter 4

I: PM state[4] entry counter 0

I: PM state[4] exit counter 0

I: PM state[5] entry counter 0

I: PM state[5] exit counter 0

I: PM state[6] entry counter 0

I: PM state[6] exit counter 0

 PASS - test_pm_singlethread in 2.466 seconds
===================================================================
TESTSUITE power_mgmt succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [power_mgmt]: pass = 3, fail = 0, skip = 0, total = 3 duration = 6.162 seconds
 - PASS - [power_mgmt.test_pm_dummyinit] duration = 1.217 seconds
 - PASS - [power_mgmt.test_pm_multithread] duration = 2.479 seconds
 - PASS - [power_mgmt.test_pm_singlethread] duration = 2.466 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

```
</details>
power consumption overview:
<img width="1452" height="508" alt="image" src="https://github.com/user-attachments/assets/b2a73676-b98a-4eae-b33f-943b0fcb287d" />

sleep mode data details:
<img width="602" height="580" alt="image" src="https://github.com/user-attachments/assets/e5d8ec77-986f-4e9b-82f8-2528b6c9e451" />

power down mode data details:
<img width="1436" height="740" alt="image" src="https://github.com/user-attachments/assets/1e753176-2b4b-414a-a492-a38ff7531fb6" />

ZhaoxiangJin committed:
dts: nxp,spc: Add wakeup-delay property for nxp spc

When a different VDD_CORE level is configured in low
power modes (LP_CFG register) compared to active modes
(ACTIVE_CFG register), the LPWKUP_DELAY register must
be configured to a minimum value.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: nxp_mcxn23x: Add power-states for MCXN23x devices

1. Add power-states for MCXN23x devices
2. Add power management related device nodes to MCXN23x,
including spc, cmc vbat, and wuu.
3. Add 'zephyr,cortex-m-idle-timer' node in frdm_mcxn236.dts

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: nxp/mcxn: Add power management support for MCXN236

Add power management support for MCXN/MCXN236.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
modules: hal_nxp: pull in spc, cmc and vbat driver for mcxn/a

pull in spc, cmc and vbat driver for mcxn/a

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: power_mgmt_soc: enable test for frdm_mcxn236

Enable power_mgmt_soc test for NXP frdm_mcxn236 board.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -91,12 +130,8 @@ static int mcux_lptmr_set_top_value(const struct device *dev,
-	flags = LPTMR_GetStatusFlags(config->base);
 
-	return ((flags & mask) == mask);
+	return LPTMR_GetStatusFlags(config->base);
```
yes, updated in [PR](https://github.com/zephyrproject-rtos/zephyr/pull/99784)

zejiang0jason commented on soc/nxp/mcx/mcxn/power.c:
```diff
@@ -0,0 +1,80 @@
+#define MCXN_CMC_ADDR		(CMC_Type *)DT_REG_ADDR(DT_INST(0, nxp_cmc))
+#define MCXN_SPC_ADDR		(SPC_Type *)DT_REG_ADDR(DT_INST(0, nxp_spc))
+
+void pm_enter_hook(void)
```
should this be `static`?

ZhaoxiangJin commented:
Update:
Most dependencies are merged, just rebased PR and addressed several comments from reviewer.
DNM tag not removed, since the last dependency got approved but not merged.
